### PR TITLE
minor changes & support json encode

### PR DESCRIPTION
**Pull request by WindzCUHK**
I tried to split the changes by commit. It maybe easier to review by commit.

@kpango 
1. The line 2157 only can pass on `go test -race`. Do you have any idea?
https://github.com/yahoojapan/athenz-authorizer/blob/2d6bbdef41652476108349d41c8439e444d0777b/policy/daemon_test.go#L2157-L2158

2. seems the retry interval is missing in jwkd: https://github.com/yahoojapan/athenz-authorizer/commit/6c79437ff203aad80ef35a9d40cdeaaffcf7b4da

3. the `ebuf` will also wrap previous error. Is this expected?
https://github.com/yahoojapan/athenz-authorizer/blob/6c79437ff203aad80ef35a9d40cdeaaffcf7b4da/jwk/daemon.go#L86-L87

### CODE REVIEW

File: jwk/daemon_test.go [hunk 1]
```diff
@@ -177,8 +177,8 @@ func Test_jwkd_Start(t *testing.T) {
 						return errors.New("cannot update keys")
 					}
 
-					if k1 == k2 {
```
Comment (kevindiu) [inline jwk/daemon_test.go]: It is more accrue to compare the pointer rather then the key ID (check if the pointer is changed or not)

Comment (WindzCUHK) [reply jwk/daemon_test.go]: new jwk.Set is created for each fetch request, so we need to check the content

Comment (kevindiu) [reply jwk/daemon_test.go]: Actually it create a new object and store into atomic.Value so I think it is ok. 
This test case is check if the j.key can be refreshed or not by the timer.

**Comment by kevindiu** [comment]
> I tried to split the changes by commit. It maybe easier to review by commit.
> 
> @kpango
> 
> 1. The line 2157 only can pass on `go test -race`. Do you have any idea?
>    https://github.com/yahoojapan/athenz-authorizer/blob/2d6bbdef41652476108349d41c8439e444d0777b/policy/daemon_test.go#L2157-L2158
> 2. seems the retry interval is missing in jwkd: [6c79437](https://github.com/yahoojapan/athenz-authorizer/commit/6c79437ff203aad80ef35a9d40cdeaaffcf7b4da)
> 3. the `ebuf` will also wrap previous error. Is this expected?
>    https://github.com/yahoojapan/athenz-authorizer/blob/6c79437ff203aad80ef35a9d40cdeaaffcf7b4da/jwk/daemon.go#L86-L87

1. Sorry I cannot reproduce the problem with `go test ./...` command.... can you paste the error message here?
2. Thank you (Y)
3. yes, it is expected....

File: jwk/daemon_test.go (hunk 1 above)
Comment (WindzCUHK) [reply jwk/daemon_test.go]: I just want to be explicit about the test case.
As updating the object is the default behavior, we cannot imply changing of the object as the actual update is done. Explicit checking for the content inside the object is required.

**Comment by WindzCUHK** [comment]
> 1. // time.Sleep(time.Duration(2) * time.Nanosecond) // not working 
>            
>    
>            
>              
>               time.Sleep(time.Duration(2) * time.Millisecond)

replace the code like this
```go
time.Sleep(time.Duration(2) * time.Nanosecond)
// time.Sleep(time.Duration(2) * time.Millisecond)
```

File: jwk/daemon_test.go (hunk 1 above)
Comment (kevindiu) [reply jwk/daemon_test.go]: It makes sense. LGTM.

**Comment by kevindiu** [comment]
The expiration of SetWithExpire() is calculated at runtime, making the expiration of the entry is not accurate, leading the object didn't delete during DeleteExpired() is called.
Gache and the policy cache do not need this level of accuracy.

File: policy/assertion.go [hunk 2]
```diff
@@ -56,3 +65,24 @@ func NewAssertion(action, resource, effect string) (*Assertion, error) {
...
+}
+
+// MarshalJSON implements the Marshaler interface.
+func (a *Assertion) MarshalJSON() ([]byte, error) {
+	x := &AssertionJSON{
+		Action:         a.Action,
+		Resource:       a.Resource,
+		ResourceDomain: a.ResourceDomain,
+		Reg:            a.Reg.String(),
+		Deny:           a.Effect == nil,
```
Comment (kevindiu) [inline policy/assertion.go]: it should be a.Effect != nil

File: policy/assertion.go [hunk 3]
```diff
@@ -56,3 +65,24 @@ func NewAssertion(action, resource, effect string) (*Assertion, error) {
 		}(),
 	}, nil
 }
+
+// AssertionJSON represents json format of Assertion
+type AssertionJSON struct {
```
Comment (kevindiu) [inline policy/assertion.go]: It shouldn't be public. Or it should be inline object for temp use in MarshalJSON function.
```suggestion
type assertionJSON struct {
```

Comment (WindzCUHK) [reply policy/assertion.go]: It is the JSON definition. There is nothing to hide.
And user need this struct to implement the decode function.

File: policy/assertion.go (hunk 2 above)
Comment (WindzCUHK) [reply policy/assertion.go]: updated

File: policy/assertion.go (hunk 3 above)
Comment (kevindiu) [reply policy/assertion.go]: It makes user confuse which object should be use, also it is internal use object for only marshaling use.

Comment (kevindiu) [reply policy/assertion.go]: Also I think this object should be 
```suggestion
type assertionJSON struct {
Assertion
RegStr string
```

Comment (WindzCUHK) [reply policy/assertion.go]: reverted

**Comment by kpango** [comment]
LGTM

**Pull request merged**